### PR TITLE
Enable jit variable bounds in dr.binary_search

### PR DESCRIPTION
**Pull request by matttsss**
This commit adds a missing functionality that is present in CPP where the binary search supports jit variable bounds.

It also adds missing typing to the function.

New tests were added as none were present for the `dr.binary_search`.

### CODE REVIEW

File: drjit/__init__.py [hunk 1]
```diff
@@ -3765,24 +3765,47 @@ def binary_search(start, end, pred):
...
     Returns:
         Index array resulting from the binary search
     '''
-    assert isinstance(start, int) and isinstance(end, int)
+    if depth_v(start) > 1 or depth_v(end) > 1 :
+        raise ValueError(f"`start` and `end` arguments should have depth <= 1, got {depth_v(start)} and {depth_v(end)}")
 
-    iterations = log2i(end - start) + 1 if start < end else 0
+    # Cast type and possibly convert to JIT the second variable
+    tp = type(start + end)
```
Comment (wjakob) [inline drjit/__init__.py]: dr.expr_t does this

File: drjit/__init__.py [hunk 2]
```diff
@@ -3765,24 +3765,47 @@ def binary_search(start, end, pred):
...
 
     Returns:
         Index array resulting from the binary search
     '''
-    assert isinstance(start, int) and isinstance(end, int)
+    if depth_v(start) > 1 or depth_v(end) > 1 :
+        raise ValueError(f"`start` and `end` arguments should have depth <= 1, got {depth_v(start)} and {depth_v(end)}")
 
-    iterations = log2i(end - start) + 1 if start < end else 0
+    # Cast type and possibly convert to JIT the second variable
```
Comment (wjakob) [inline drjit/__init__.py]: > # Cast type and possibly convert to JIT the second variable
Ungrammatical.

I would say: "Transform inputs to a consistent type" or similar

File: drjit/__init__.py [hunk 3]
```diff
@@ -3765,24 +3765,47 @@ def binary_search(start, end, pred):
...
+    tp = type(start + end)
+    start = tp(start)
+    end = tp(end)
 
-    for _ in range(iterations):
+    iterations = detach(select(start < end, log2i(abs(end - start)) + 1, 0))
```
Comment (wjakob) [inline drjit/__init__.py]: I don't understand the `detach`. These are not differentiable variables, and it should not matter whether we access them through `.ad` types or not.

File: tests/test_binary_search.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
+import drjit as dr
+import numpy as np
+import pytest
+
+
+def _ref(start, end, values, thresholds):
```
Comment (wjakob) [inline tests/test_binary_search.py:6]: This kind of test coverage is a bit insane and some of the tests look highly redundant. Is it by any chance AI-generated?

Please use parameterized tests to cover more variants with less code. We don't need many individual tests for obscure corner cases. Often some of them can be rolled into 1 extra line of a combined test.

File: drjit/__init__.py (hunk 1 above)
Comment (matttsss) [reply drjit/__init__.py]: I wasn't aware, thanks!

File: drjit/__init__.py (hunk 3 above)
Comment (matttsss) [reply drjit/__init__.py]: There is no reason yes, I should have removed it in the cleanup

File: tests/test_binary_search.py (hunk 4 above)
Comment (matttsss) [reply tests/test_binary_search.py:6]: Yes they are, I already did a cleanup pass on them but I agree they were still a bit long
